### ISSUE TITLE
[Masonry] Implement unified PointerEvent

### ISSUE THREAD

**Issue by xorgy** [linebender/xilem#312]
There is some consensus that a PointerEvent-style API (like existed in Glazier) is probably a good way to handle touch and mouse in general. This is a tracking issue for that topic.

**Comment by raphlinus** [comment]
There are a few layers to this. The simplest layer is just to enable mouse-like behavior from touch, so that touch can operate widgets like buttons and sliders, and also selection (and focus) in text input.

Another more sophisticated layer is to support multi-touch well. That has potentially far-reaching changes, as multiple widgets can be active at the same time. A careful design is needed.

And yet another more sophisticated layer is gesture disambiguation and recognition. A touch-drag gesture on a slider widget should move the slider, but on mobile, the same gesture *not* on a suitably sensitive widget should be interpreted as a scroll. There *may* be platform-specific gesture recognition code we want to call into, or it may be that we want a pure Rust cross-platform implementation of all this, tunable to feel native on the specific platform.

There may also be improvements desired on the winit side. I see a bunch of PRs to add pen support (https://github.com/rust-windowing/winit/pull/2647 is the most recent) but they seem to be stuck in review. https://github.com/rust-windowing/winit/issues/99 seems to be the main tracking issue there.

To be clear, I think we should have the simplest layer first, so we can unblock simple UI interactions on mobile, then carefully design what we want to do for the other two.

**Comment by PoignardAzur** [comment]
How much progress have we made on the Android side of this?

**Comment by xorgy** [comment]
@PoignardAzur 
Winit has unified their PointerEvent interface upstream, and 0.31.0 will have this, they have also slated click counts on winit `PointerEvent`s for 0.31.0.
I think at this point it'd be best to just keep what we have, and when we move to winit 0.31.0 we can make things more elegant (and delete a bunch of code).

We still have to work out properly intercepting PointerEvents from touch (and maybe pen) devices. On the web, the UA handles this under the hood. Handling touch gesture recognition properly will still involve retaining the full history of a touch (timestamp, position, and force) between `Start` and `End`/`Cancel`

**Comment by PoignardAzur** [comment]
I think we've implemented all of the above.

**Comment by DJMcNab** [comment]
I'm not sure that we have. Our handling of touch is currently to pretend that they are the mouse, which isn't really right.

**Comment by xorgy** [comment]
> I'm not sure that we have. Our handling of touch is currently to pretend that they are the mouse, which isn't really right.

Yeah, in touch environments there needs to be a way to give containers a bite at a touch before their children, and if you want to handle pens in the expected way on windows, you also need this for pen inputs.

**Comment by xorgy** [comment]
This issue has been explored more thoroughly in the ui-events space, and I think it's fair to simply close the issue. PointerEvent is roughly as unified as it can practically be on the platforms supported, and whenever possible converges toward this value.
